java/tools/src/java/org/apache/orc/tools/ScanData.java

```diff
@@ -40,41 +40,168 @@
+    } else {
+      long rowInStripe = current.row + batchSize - current.firstRow;
+      result = current.firstRow + (rowInStripe + stride - 1) / stride * stride;
+    }
+    return findStripeInfo(reader, result);
+  }
+
+  static boolean findBadColumns(Reader reader, LocationInfo current, int batchSize,
+      TypeDescription column, boolean[] include) {
+    include[column.getId()] = true;
+    TypeDescription schema = reader.getSchema();
+    boolean result = false;
+    if (column.getChildren() == null) {
+      int row = 0;
+      try (RecordReader rows = reader.rows(reader.options().include(include))) {
+        rows.seekToRow(current.row);
+        VectorizedRowBatch batch = schema.createRowBatch(
+            TypeDescription.RowBatchVersion.USE_DECIMAL64, 1);
+        for(row=0; row < batchSize; ++row) {
+          rows.nextBatch(batch);
+        }
+      } catch (Throwable t) {
+        System.out.printf("Column %d failed at row %d%n", column.getId(),
+            current.row + row);
+        result = true;
+      }
+    } else {
+      for(TypeDescription child: column.getChildren()) {
+        result |= findBadColumns(reader, current, batchSize, child, include);
+      }
+    }
+    include[column.getId()] = false;
+    return result;
+  }
+
+  static void main(Configuration conf, String[] args) throws ParseException {
     CommandLine cli = parseCommandLine(args);
     if (cli.hasOption('h') || cli.getArgs().length == 0) {
-      System.err.println("usage: java -jar orc-tools-*.jar scan [--help] <orc file>*");
+      System.err.println("usage: java -jar orc-tools-*.jar scan [--schema] [--help] <orc file>*");
       System.exit(1);
     } else {
+      boolean printSchema = cli.hasOption('s');
       List<String> badFiles = new ArrayList<>();
       for (String file : cli.getArgs()) {
-        try {
-          Path path = new Path(file);
-          Reader reader = FileDump.getReader(path, conf, badFiles);
-          if (reader == null) {
-            continue;
-          }
-          RecordReader rows = reader.rows();
-          VectorizedRowBatch batch = reader.getSchema().createRowBatch();
-          long batchCount = 0;
-          long rowCount = 0;
-          while (rows.nextBatch(batch)) {
-            batchCount += 1;
-            rowCount += batch.size;
+        try (Reader reader = FileDump.getReader(new Path(file), conf, badFiles)) {
+          if (reader != null) {
+            TypeDescription schema = reader.getSchema();
+            if (printSchema) {
+              System.out.println(schema.toJson());
+            }
+            VectorizedRowBatch batch = schema.createRowBatch(
+                TypeDescription.RowBatchVersion.USE_DECIMAL64,
+                calculateBestVectorSize(reader.getRowIndexStride()));
+            final int batchSize = batch.getMaxSize();
+            long badBatches = 0;
+            long currentRow = 0;
+            long goodRows = 0;
+            try (RecordReader rows = reader.rows()) {
```

Comment:
Can we handle the corrupted file more gracefully? This may be considered as a regression in the following case.

**BEFORE**
```
$ orc-tools scan ../examples/corrupt/stripe_footer_bad_column_encodings.orc
log4j:WARN No appenders could be found for logger (org.apache.hadoop.util.Shell).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Processing data file ../examples/corrupt/stripe_footer_bad_column_encodings.orc [length: 780]
Unable to dump data for file: ../examples/corrupt/stripe_footer_bad_column_encodings.orc
```

**AFTER (this PR)**
```
$ java -jar tools/target/orc-tools-1.7.0-SNAPSHOT-uber.jar scan ../examples/corrupt/stripe_footer_bad_column_encodings.orc
log4j:WARN No appenders could be found for logger (org.apache.hadoop.util.Shell).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Processing data file ../examples/corrupt/stripe_footer_bad_column_encodings.orc [length: 780]
Unable to open file: ../examples/corrupt/stripe_footer_bad_column_encodings.orc
java.lang.IndexOutOfBoundsException: Index: 0
	at java.util.Collections$EmptyList.get(Collections.java:4456)
	at org.apache.orc.OrcProto$StripeFooter.getColumns(OrcProto.java:14080)
	at org.apache.orc.impl.reader.StripePlanner.buildEncodings(StripePlanner.java:224)
	at org.apache.orc.impl.reader.StripePlanner.parseStripe(StripePlanner.java:126)
	at org.apache.orc.impl.RecordReaderImpl.readStripe(RecordReaderImpl.java:1117)
	at org.apache.orc.impl.RecordReaderImpl.advanceStripe(RecordReaderImpl.java:1168)
	at org.apache.orc.impl.RecordReaderImpl.advanceToNextRow(RecordReaderImpl.java:1203)
	at org.apache.orc.impl.RecordReaderImpl.<init>(RecordReaderImpl.java:268)
	at org.apache.orc.impl.ReaderImpl.rows(ReaderImpl.java:841)
	at org.apache.orc.impl.ReaderImpl.rows(ReaderImpl.java:835)
	at org.apache.orc.tools.ScanData.main(ScanData.java:171)
	at org.apache.orc.tools.Driver.main(Driver.java:126)
```

Reply:
In this case, instead of showing the raw exception like `IndexOutOfBoundsException` directly, can we say about `corrupted footer` as a general message?

Reply:
Yes, although the Scan tool doesn't have the information about where problems happen, when it is the footer. We'd need to add better exceptions in the ReaderImpl, which is a good idea.

Reply:
Do you want me to hide the actual exceptions behind a "-v" option?

Reply:
Oh, `-v` sounds much better to me. Thanks!